aws-storage-s3/src/androidTest/java/com/amplifyframework/storage/s3/AWSS3StorageUploadTest.java

```diff
@@ -0,0 +1,238 @@
+        latchedUploadAndConfirm(largeFile, DEFAULT_ACCESS_LEVEL, getIdentityId());
+    }
+
+    /**
+     * Tests that file upload operation can be canceled while the
+     * transfer hasn't completed yet.
+     *
+     * @throws Exception if upload is not canceled successfully
+     *         before timeout
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUploadFileIsCancelable() throws Exception {
+        final CountDownLatch canceled = new CountDownLatch(1);
+
+        // Begin uploading large file
+        StorageUploadFileOperation<?> op = Amplify.Storage.uploadFile(
+            largeFile.getName(),
+            largeFile.getAbsolutePath(),
+            options,
+            onResult -> fail("Upload finished before being successfully cancelled."),
+            onError -> fail("Upload failed for a different reason.")
+        );
+
+        // Listen to Hub events to cancel when progress has been made
+        SubscriptionToken progressToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadProgress".equals(hubEvent.getName())) {
+                HubEvent<Float> progressEvent = (HubEvent<Float>) hubEvent;
+                Float progress = progressEvent.getData();
+                if (progress != null && progress > 0) {
+                    op.cancel();
+                }
+            }
+        });
+        subscriptions.add(progressToken);
+
+        // Listen to Hub events for cancel
+        SubscriptionToken cancelToken = Amplify.Hub.subscribe(HubChannel.STORAGE, hubEvent -> {
+            if ("uploadState".equals(hubEvent.getName())) {
+                HubEvent<String> stateEvent = (HubEvent<String>) hubEvent;
+                TransferState state = TransferState.getState(stateEvent.getData());
+                if (TransferState.CANCELED.equals(state)) {
+                    canceled.countDown();
+                }
+            }
+        });
+        subscriptions.add(cancelToken);
+
+        // Assert that the required conditions have been met
+        assertTrue(canceled.await(EXTENDED_TIMEOUT_IN_SECONDS, TimeUnit.SECONDS));
+    }
+
+    /**
+     * Tests that file upload operation can be paused and resumed
+     * while the transfer hasn't completed yet.
+     *
+     * @throws Exception if upload is not paused, resumed, and
+     *         completed successfully before timeout
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testUploadFileIsResumable() throws Exception {
+        final CountDownLatch completed = new CountDownLatch(1);
+        final CountDownLatch resumed = new CountDownLatch(1);
+
+        // Begin uploading large file
+        StorageUploadFileOperation<?> op = Amplify.Storage.uploadFile(
+            largeFile.getName(),
+            largeFile.getAbsolutePath(),
+            options,
+            onResult -> completed.countDown(),
+            onError -> fail("Upload is not successful.")
```

Comment:
This probably isn't going to work. This will crash the thread on which the callback happens. But it _won't_ crash the jUnit thread. The `completed.countDown()` approach does work.

But, this whole thing can be replaced by my `Await` utility (this is 100% what that class is for.)

If you put `uploadFile` into a `SynchronousStorage` utility, then have:
```
    StorageUploadFileResult uploadFile(String name, String path, StorageOptions options) throws StorageException {
    return Await.result((onResult, onError) -> {
        Amplify.Storage.uploadFile(..., onResult, onError);
    });
}
```

Reply:
I intentionally didn't use `Await` because the code needs to be unblocked. If you look closely, you will notice that I am not using the result, but the actual storage operation so that I can pause and resume it.

Reply:
Oops, sorry. Sounds good!